### DISCUSSION TITLE
SAML

### DISCUSSION THREAD

**Discussion by tumbl3w33d** [kanidm/kanidm#2874] [General]
I would like to gather some input of knowledgeable folks on the SAML topic. Currently kanidm only supports OIDC, which is fine for a lot of use cases. Yet there are numerous others, especially in the enterprise universe, which only only support SAML.

There is a [ticket](https://github.com/kanidm/kanidm/issues/2025) on the roadmap to support SAML in kani, however, it was recently postponed from what I saw and it is totally fine for me and I don't mean to produce any stress with this discussion.

I am more interested in the existing options that I have to workaround this limitation. There are several kinds of adapters / bridges between OIDC and SAML. @yaleman mentioned https://simplesamlphp.org/ and I found [SATOSA](https://github.com/IdentityPython/SATOSA). So, while I understand OIDC to a certain extent, I have no experience with SAML other than hearing about its complexity. I would like to hear your educated opinions about such bridge approaches regarding security implications, i.e., how much can go wrong.

I would also like to know, if it is apparently so "easy" to bridge from A to B, why we don't add such proxy functionality to kani (or is that what the mentioned ticket would do?). It might also make sense to add a note to kani's faq, because the need seems not so exotic. My latest use case was using kani with AWS, which seems possible but is far from a low-hanging fruit without SAML support.

Looking forward to your input!

**Comment by yaleman** [comment]
There's no simple option, only half-secured hacks at this time. SimpleSAMLPHP I've only ever been able to get working with LDAP, which means single-factor auth only. 

The README for SATOSA looks curious, I'd be curious to see how it works in practice.

SAML's a VERY complex protocol to get right, with an incredibly flexible (and complex) set of options and MANY ways to build insecure environments. I haven't found any XML canonicalisation or signing libraries for Rust yet, after spending quite some time looking... otherwise I'd have taken up working on [terminaloutcomes/saml-rs](https://github.com/terminaloutcomes/saml-rs) which I had to give up on a few years back 😢

**Reply by tumbl3w33d** [reply]
The realization that there simply is no way to strap this on top of an OIDC IDP without compromising security would also be a good answer and save me and others from from exploring options further. :)

**Reply by yaleman** [reply]
I haven't found one yet, but in this discussion you've already highlighted another package which *might* work - someone might be able to indicate a better one!

But yes, at this point, the only options I've heard of have serious compromises. I'd gladly hear about better ones!

**Comment by yaleman** [comment]
As an aside, Kanidm is not an authentication proxy. It's an Identity provider. There's a big difference.

**Reply by tumbl3w33d** [reply]
I think I understood that part, however, I don't see a contradiction to why it cannot proxy to itself, either with a separate project or a feature added to the main stack, both to be activated like the current LDAP interface.

**Reply by yaleman** [reply]
This is the clarity I'm trying to build... 

> why it cannot proxy to itself

There's no "proxying" in Kanidm. Kanidm *implements* the OIDC and LDAP protocols, it doesn't "proxy" to anything. 

An "Authentication proxy" is something like Apache, Nginx or oauth2-proxy, that sits in the network path between the application and the user.

> with a separate project

This would be using SimpleSAMLPHP or another, which you log into and it contacts Kanidm via LDAP or another protocol to confirm user data and/or authentication.

>  a feature added to the main stack,

This would be the "implement SAML protocol support" part, which requires someone to find a way to do it...

**Reply by Firstyear** [reply]
The whole auth space has been filled with a bunch of thought leaders who throw around words like "proxy" and "broker" and "federation" without actually defining what they mean. Don't even start me on decentralised .... It's really done us a huge disservice in the space.  

Anyway, in this case using SimpleSAMLPhp would be the best way to go. Simple SAML PHP would then direct to oauth/oidc on Kanidm, and if that succeeds, SimpleSAMLPHP then does the SAML side.

**Comment by jinnatar** [comment]
Thanks for the mention of SATOSA! I did a deep dive and came up with a guide and a proof of concept proxy container image that even has PKCE: https://github.com/jinnatar/satosa-saml-proxy

I've been so deep in it for the past two days I might not be seeing something being wrong so feedback is much appreciated. I will also try to improve upstream documentation since it was a bit lacking especially for PKCE.

**Reply by Firstyear** [reply]
I think @yaleman Has the best insight here to review this :)
